File: test/extensions/dynamic_modules/test_data/go/load_balancer_integration_test/load_balancer_integration_test.go
// Load balancer integration test module.
//
// Registers a "first_host_lb" load balancer that always picks priority 0, index 0 (the
// only host in the test setup). The test asserts that requests successfully route
// through Envoy with the dynamic-module LB attached, exercising the full
// ChooseHost / OnHostMembershipUpdate dispatch surface.
package main

import (
	sdk "github.com/envoyproxy/envoy/source/extensions/dynamic_modules/sdk/go"
	_ "github.com/envoyproxy/envoy/source/extensions/dynamic_modules/sdk/go/abi"
	"github.com/envoyproxy/envoy/source/extensions/dynamic_modules/sdk/go/shared"
)

func init() {
	sdk.RegisterLoadBalancerConfigFactories(map[string]shared.LoadBalancerConfigFactory{
		"first_host_lb": &firstHostConfigFactory{},
	})
}

func main() {} //nolint:all

type firstHostConfigFactory struct {
	shared.EmptyLoadBalancerConfigFactory
}

func (firstHostConfigFactory) Create(_ shared.LoadBalancerConfigHandle, _ []byte) (shared.LoadBalancerFactory, error) {
	return &firstHostFactory{}, nil
}

type firstHostFactory struct {
	shared.EmptyLoadBalancerFactory
}

func (*firstHostFactory) Create(_ shared.LoadBalancerHandle) shared.LoadBalancer {
	return &firstHostLB{}
}

type firstHostLB struct {
	shared.EmptyLoadBalancer
}

func (*firstHostLB) ChooseHost(handle shared.LoadBalancerHandle, _ shared.LoadBalancerContext) (shared.HostSelection, bool) {
	// Bail if no hosts are available.
	if handle.GetHostsCount(0) == 0 {
		return shared.HostSelection{}, false
	}
	return shared.HostSelection{Priority: 0, Index: 0}, true
}

File: test/extensions/dynamic_modules/test_data/rust/load_balancer_integration_test.rs
//! Load balancer integration test module. Mirror of
//! test_data/go/load_balancer_integration_test/load_balancer_integration_test.go.
//!
//! Registers a "first_host_lb" load balancer that always picks priority 0, index 0.

use envoy_proxy_dynamic_modules_rust_sdk::*;
use std::sync::Arc;

declare_load_balancer_init_functions!(init, new_lb_config);

fn init() -> bool {
  true
}

fn new_lb_config(
  _name: &str,
  _config: &[u8],
  _envoy_lb_config: Arc<dyn EnvoyLbConfig>,
) -> Option<Box<dyn LoadBalancerConfig>> {
  Some(Box::new(FirstHostConfig {}))
}

struct FirstHostConfig {}

impl LoadBalancerConfig for FirstHostConfig {
  fn new_load_balancer(&self, _envoy_lb: &dyn EnvoyLoadBalancer) -> Box<dyn LoadBalancer> {
    Box::new(FirstHostLB {})
  }
}

struct FirstHostLB {}

impl LoadBalancer for FirstHostLB {
  fn choose_host(&mut self, envoy_lb: &dyn EnvoyLoadBalancer) -> Option<HostSelection> {
    if envoy_lb.get_hosts_count(0) == 0 {
      return None;
    }
    Some(HostSelection {
      priority: 0,
      index: 0,
    })
  }
}
